### DISCUSSION TITLE
Providers and DAOs

### DISCUSSION THREAD

**Discussion by ghost** [simolus3/drift#1528] [Q&A]
I am struggling to get DAO's to work, more my lack of knowledge than anything else, here is my code (just testing stuff at the moment).

void main() {
  final db = MyDb();
  runApp(
    Provider<MyDb>(
      create: (context) => db,
      child: MyApp(),
    ),
  );
}

the above works if you make calls like

`var db = Provider.of<MyDb>(context, listen: false);
`
inside widgets etc, where I am struggling is using the provider model I was expecting the following to work.

`var dao = Provider.of<TblJobsDao>(context, listen: false);
`
it fails with the following error

======== Exception caught by gesture ===============================================================
The following ProviderNotFoundException was thrown while handling a gesture:
Error: Could not find the correct Provider<TblJobsDao> above this MyHomePage Widget

This happens because you used a `BuildContext` that does not include the provider
of your choice. There are a few common scenarios:

- You added a new provider in your `main.dart` and performed a hot-reload.
  To fix, perform a hot-restart.

- The provider you are trying to read is in a different route.

  Providers are "scoped". So if you insert of provider inside a route, then
  other routes will not be able to access that provider.

- You used a `BuildContext` that is an ancestor of the provider you are trying to read.

  Make sure that MyHomePage is under your MultiProvider/Provider<TblJobsDao>.
  This usually happens when you are creating a provider and trying to read it immediately.

  For example, instead of:

  ```
  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // Will throw a ProviderNotFoundError, because `context` is associated
      // to the widget that is the parent of `Provider<Example>`
      child: Text(context.watch<Example>()),
    ),
  }
  ```

  consider using `builder` like so:

  ```
  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // we use `builder` to obtain a new `BuildContext` that has access to the provider
      builder: (context) {
        // No longer throws
        return Text(context.watch<Example>()),
      }
    ),
  }

I am guessing it's how the provider is setup?

Now I am being lazy, what would be nice is a basic example of provider based solution using DAO's I am guessing this is used a lot once you have a lot of tables.

Thank you.

**Comment by AlexW68** [comment]
Thanks for the reply, made the changes

```
void main() {
  runApp(
    Provider(
      create: (context) => MyDb(),
      child: MyApp(),
    ),
  );
}
```

And this in the default app

```
  void _incrementCounter() {
    final db = Provider.of<MyDb>(context, listen: false);
    final job = TblJobsCompanion(startAddress: Value("1 road"));
    db.insertTblJob(job);
  }
```
works perfectly but I cannot work out how to reference the dao

so this code fails (with the error in the original post)

```
  void _incrementCounter() {
    var dao = Provider.of<TblJobsDao>(context, listen: false);
    final job = TblJobsCompanion(startAddress: Value("1 road"));
    dao.insertTblJob(job);
  }

```

I am missing a piece of the puzzle somewhere.

the tbljobs_dao.g.dart file might not have the right info in it, here is the contents.

```
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'tbljob_dao.dart';

// **************************************************************************
// DaoGenerator
// **************************************************************************

mixin _$TblJobDaoMixin on DatabaseAccessor<MyDb> {}

```

Will look at Riverpod once I have something working as there is nothing worse than trying to learn two related things at the same time.

Thank you.

**Comment by North101** [comment]
It fails with `Provider.of<TblJobsDao>(context);` because you haven't told `Provider` what `TblJobsDao` is or how to create it. Try this (untested):

```dart
void main() {
  return runApp(
    MultiProvider(
      providers: [
        Provider<MyDb>(
          create: (context) => MyDb(),
          dispose: (context, db) => db.close(),
        ),
        Provider<TblJobsDao>(
          create: (context) {
            final db = Provider.of<MyDb>(context);
            return TblJobsDao(db);
          },
        ),
      ],
      child: MyApp(),
    ),
  );
}
```

**Reply by AlexW68** [reply]
Thank you for the reply, you got me 99% of the way there (my lack of understanding more than anything).

Here is what I ended up with

```
void main() {
  return runApp(
    MultiProvider(
      providers: [
        Provider<MyDb>(
          create: (context) => MyDb(),
          dispose: (context, db) => db.close(),
        ),
        Provider<TblJobsDao>(
          create: (context) {
            final db = Provider.of<MyDb>(context, listen: false);
            return TblJobsDao(db);
          },
        ),
      ],
      child: MyApp(),
    ),
  );
}
```

Without  the listen: false another failure happens, hot reloads, hot restarts don't work when testing, complete restarts when changing the initial provider configuration is the only way I got things to work.

If I leave out the listen: false I get the following error I am putting  it here just in case anyone else hits the same problem

Bad state: Tried to read a provider that threw during the creation of its value.
The exception occurred during the creation of type TblJobsDao.

Thanks again.
